sql/core/src/main/scala/org/apache/spark/sql/execution/GenerateExec.scala

```diff
@@ -40,6 +42,10 @@ private[execution] sealed case class LazyIterator(func: () => TraversableOnce[In
  * output of each into a new stream of rows.  This operation is similar to a `flatMap` in functional
  * programming with one important additional feature, which allows the input rows to be joined with
  * their output.
+ *
+ * This operator supports whole stage code generation for generators that do not implement
+ * terminate().
```

Comment:
I don't think this holds. This method implies that the generator holds some state, and that calling this on an un-used generator might return the correct result.

I have added an explicit `GeneratorWithCodegen` trait which all generators must implement if they want to be eligible for code generation. The terminate() method in this trait has been closed off.